bugsnag-plugin-android-anr/src/main/jni/anr_handler.c

```diff
@@ -5,118 +5,128 @@
 #include <pthread.h>
 #include <signal.h>
 #include <string.h>
+#include <unistd.h>
 
-#include "utils/string.h"
-
-static pthread_t bsg_watchdog_thread;
-static sigset_t bsg_anr_sigmask;
+#include "anr_google.h"
 
 // Lock for changing the handler configuration
 static pthread_mutex_t bsg_anr_handler_config = PTHREAD_MUTEX_INITIALIZER;
+
 // A proxy for install/uninstall state, to avoid needing to unset the handler
 // on the sigquit-watching thread
 static bool enabled = false;
 static bool installed = false;
-static bool invokePrevHandler = true;
+
+static pthread_t watchdog_thread;
+static volatile bool should_report_anr = false;
+static struct sigaction original_sigquit_handler;
 
 static JavaVM *bsg_jvm = NULL;
 static jmethodID mthd_notify_anr_detected = NULL;
 static jobject obj_plugin = NULL;
 
-/* The previous SIGQUIT handler */
-struct sigaction bsg_sigquit_sigaction_previous;
-
-void bsg_notify_anr_detected(JNIEnv *env) {
-  (*env)->CallVoidMethod(env, obj_plugin, mthd_notify_anr_detected);
-}
-
-bool bsg_configure_anr_jni(JNIEnv *env) {
+static bool configure_anr_jni(JNIEnv *env) {
   // get a global reference to the AnrPlugin class
   // https://developer.android.com/training/articles/perf-jni#faq:-why-didnt-findclass-find-my-class
   int result = (*env)->GetJavaVM(env, &bsg_jvm);
   if (result != 0) {
-    // Failed to fetch VM, cannot continue
+    BUGSNAG_LOG("Failed to fetch Java VM. ANR handler not installed.");
     return false;
   }
 
   jclass clz = (*env)->FindClass(env, "com/bugsnag/android/AnrPlugin");
   mthd_notify_anr_detected =
-      (*env)->GetMethodID(env, clz, "notifyAnrDetected", "()V");
+          (*env)->GetMethodID(env, clz, "notifyAnrDetected", "()V");
   return true;
 }
 
-void bsg_handle_sigquit(int signum, siginfo_t *info, void *user_context) {
+static void notify_anr_detected() {
   if (enabled) {
-    // invoke a JNI call from the SIGQUIT handler
     JNIEnv *env;
-    int result = (*bsg_jvm)->GetEnv(bsg_jvm, (void **)&env, JNI_VERSION_1_4);
+    int result = (*bsg_jvm)->GetEnv(bsg_jvm, (void **) &env, JNI_VERSION_1_4);
 
     if (result == JNI_OK) { // already attached
-      bsg_notify_anr_detected(env);
+      (*env)->CallVoidMethod(env, obj_plugin, mthd_notify_anr_detected);
```

Comment:
This is dead code (and illegal if it ever becomes live).

Reply:
Could you please explain more about what's wrong? This is following recommendations from https://stackoverflow.com/a/12901159 to handle all edge cases of the environment's attached state.

Reply:
As I said in bug https://github.com/bugsnag/bugsnag-android/issues/1026 attempting to make any JNI calls from a signal handler when the thread was running java or runtime code can put the runtime into an illegal state.

Reply:
Yes, thanks for that, we took it to heart! This code gets called from the thread running `sigquit_watchdog_thread_main`, which waits on `should_report_anr` and then calls `notify_anr_detected()`. The SIGQUIT handler sets `should_report_anr` to true and then exits (primitive, but async-safe). So no Java code is called from the signal handler's thread. Unfortunately the diff here is horribly messy; the finished code is a lot easier to follow: https://github.com/bugsnag/bugsnag-android/blob/1552f7d58396fca8b99630516487641a901bfb7a/bugsnag-plugin-android-anr/src/main/jni/anr_handler.c#L43